src/OrleansAzureUtils/Providers/Storage/AzureBlobStorage.cs

```diff
@@ -0,0 +1,235 @@
+using Microsoft.WindowsAzure.Storage.Blob.Protocol;
+using Newtonsoft.Json;
+using Orleans.Providers;
+using Orleans.Runtime;
+
+namespace Orleans.Storage
+{
+    /// <summary>
+    /// Simple storage provider for writing grain state data to Azure blob storage in JSON format.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Required configuration params: <c>DataConnectionString</c>
+    /// </para>
+    /// <para>
+    /// Optional configuration params:
+    /// <c>ContainerName</c> -- defaults to <c>grainstate</c>
+    /// <c>SerializeTypeNames</c> -- defaults to <c>OrleansGrainState</c>
+    /// <c>PreserveReferencesHandling</c> -- defaults to <c>false</c>
+    /// <c>UseFullAssemblyNames</c> -- defaults to <c>false</c>
+    /// <c>IndentJSON</c> -- defaults to <c>false</c>
+    /// </para>
+    /// </remarks>
+    /// <example>
+    /// Example configuration for this storage provider in OrleansConfiguration.xml file:
+    /// <code>
+    /// &lt;OrleansConfiguration xmlns="urn:orleans">
+    ///   &lt;Globals>
+    ///     &lt;StorageProviders>
+    ///       &lt;Provider Type="Orleans.Storage.AzureBlobStorage" Name="AzureStore"
+    ///         DataConnectionString="UseDevelopmentStorage=true"
+    ///       />
+    ///   &lt;/StorageProviders>
+    /// </code>
+    /// </example>
+    public class AzureBlobStorage : IStorageProvider
+    {
+        private JsonSerializerSettings settings;
+
+        private CloudBlobContainer container;
+
+        /// <summary> Logger used by this storage provider instance. </summary>
+        /// <see cref="IStorageProvider.Log"/>
+        public Logger Log { get; private set; }
+
+        /// <summary> Name of this storage provider instance. </summary>
+        /// <see cref="IProvider.Name"/>
+        public string Name { get; private set; }
+
+        /// <summary> Initialization function for this storage provider. </summary>
+        /// <see cref="IProvider.Init"/>
+        public async Task Init(string name, IProviderRuntime providerRuntime, IProviderConfiguration config)
+        {
+            Log = providerRuntime.GetLogger(this.GetType().Name);
+
+            try
+            {
+                ConfigureJsonSerializerSettings(config);
+
+                if (!config.Properties.ContainsKey("DataConnectionString")) throw new BadProviderConfigException("The DataConnectionString setting has not been configured in the cloud role. Please add a DataConnectionString setting with a valid Azure Storage connection string.");
+
+                var account = CloudStorageAccount.Parse(config.Properties["DataConnectionString"]);
+                var blobClient = account.CreateCloudBlobClient();
+                var containerName = config.Properties.ContainsKey("ContainerName") ? config.Properties["ContainerName"] : "grainstate";
+                container = blobClient.GetContainerReference(containerName);
+                await container.CreateIfNotExistsAsync();
```

Comment:
Ideally, I wouldn't try to create the container at init and then never again. I would instead (or in addition, if you want to have this call effectively to check access), whenever you are writing a blob and you get a ContainerNotFound error, then create it and retry the write... This simplifies messing with the blobs at runtime and not getting the silos into an irrecoverable state unless they are restarted.